controllers/podchaos/podnotready/types.go

```diff
@@ -0,0 +1,349 @@
+	"k8s.io/apimachinery/pkg/util/wait"
+	"k8s.io/client-go/tools/cache"
+	ctrl "sigs.k8s.io/controller-runtime"
+
+	"github.com/chaos-mesh/chaos-mesh/pkg/router"
+	ctx "github.com/chaos-mesh/chaos-mesh/pkg/router/context"
+	end "github.com/chaos-mesh/chaos-mesh/pkg/router/endpoint"
+
+	"github.com/chaos-mesh/chaos-mesh/api/v1alpha1"
+	"github.com/chaos-mesh/chaos-mesh/controllers/common"
+	"github.com/chaos-mesh/chaos-mesh/pkg/utils"
+)
+
+const (
+	podNotReadyActionMsg = "pod notready duration %s"
+
+	ChaosMeshInjectNotReady v1.PodConditionType = "ChaosMeshInjectNotReady"
+)
+
+type endpoint struct {
+	ctx.Context
+}
+
+// Object implements the reconciler.InnerReconciler.Object
+func (r *endpoint) Object() v1alpha1.InnerObject {
+	return &v1alpha1.PodChaos{}
+}
+
+// Apply implements the reconciler.InnerReconciler.Apply
+func (r *endpoint) Apply(ctx context.Context, req ctrl.Request, chaos v1alpha1.InnerObject) error {
+
+	podchaos, ok := chaos.(*v1alpha1.PodChaos)
+	if !ok {
+		err := errors.New("chaos is not PodChaos")
+		r.Log.Error(err, "chaos is not PodChaos", "chaos", chaos)
+		return err
+	}
+
+	pods, err := utils.SelectAndFilterPods(ctx, r.Client, r.Reader, &podchaos.Spec)
+	if err != nil {
+		r.Log.Error(err, "failed to select and filter pods")
+		return err
+	}
+	err = r.setAllPodsNotReady(ctx, pods, podchaos)
+	if err != nil {
+		return err
+	}
+
+	podchaos.Status.Experiment.PodRecords = make([]v1alpha1.PodStatus, 0, len(pods))
+	for _, pod := range pods {
+		ps := v1alpha1.PodStatus{
+			Namespace: pod.Namespace,
+			Name:      pod.Name,
+			HostIP:    pod.Status.HostIP,
+			PodIP:     pod.Status.PodIP,
+			Action:    string(podchaos.Spec.Action),
+		}
+		if podchaos.Spec.Duration != nil {
+			ps.Message = fmt.Sprintf(podNotReadyActionMsg, *podchaos.Spec.Duration)
+		}
+		podchaos.Status.Experiment.PodRecords = append(podchaos.Status.Experiment.PodRecords, ps)
+	}
+	r.Event(podchaos, v1.EventTypeNormal, utils.EventChaosInjected, "")
+	return nil
+}
+
+func (r *endpoint) setAllPodsNotReady(ctx context.Context, pods []v1.Pod, podchaos *v1alpha1.PodChaos) error {
+	for index := range pods {
+		pod := &pods[index]
+		if containsReadinessGate(pod) {
+			continue
+		}
+		key, err := cache.MetaNamespaceKeyFunc(pod)
+		if err != nil {
+			return err
+		}
+		podchaos.Finalizers = utils.InsertFinalizer(podchaos.Finalizers, key)
+
+		clonePod := &v1.Pod{
+			ObjectMeta: metav1.ObjectMeta{
+				Name:            pod.Name,
+				Labels:          pod.Labels,
+				Annotations:     pod.Annotations,
+				GenerateName:    pod.GenerateName,
+				Finalizers:      pod.Finalizers,
+				OwnerReferences: pod.OwnerReferences,
+			},
+		}
+		if len(pod.Namespace) > 0 {
+			clonePod.ObjectMeta.Namespace = pod.Namespace
+		}
+		clonePod.Spec = *pod.Spec.DeepCopy()
+
+		r.Log.Info("Add readinessGate spec into pod", "namespace", pod.Namespace, "name", pod.Name)
+		injectReadinessGate(clonePod)
+
+		if err := r.Delete(ctx, pod); err != nil {
```

Comment:
We deleted the POD and then rebuilt it. What happens if the POD is controlled by Deployment or StatefulSet?

Reply:
It's uncontrollable. Without deleting the pod, do you have a better way？

Reply:
What difference between `pod-failure` and this action?

Reply:
Can we use the admission-webhook to interfere with the pod creation process?  There will be more work to be done in using `admission-webhook`.